packages/cli/src/modules/auth/commands/login.ts

```diff
@@ -0,0 +1,364 @@
+  authorize.searchParams.set('code_challenge_method', 'S256');
+  return authorize.toString();
+}
+
+async function openBrowserOrPrint(url: string, noBrowser?: boolean) {
+  if (noBrowser) {
+    process.stderr.write(`Open this URL to continue: ${url}\n`);
+  } else {
+    process.stderr.write(`Opening the following URL: ${url}\n`);
+    openInBrowser(url);
+  }
+}
+
+async function waitForAuthorizationCode(
+  callback: Awaited<ReturnType<typeof startCallbackServer>>,
+  expectedState: string,
+) {
+  const { code, state } = await callback.waitForCode();
+  await callback.close();
+  if (state !== expectedState) {
+    throw new Error('State mismatch');
+  }
+  return code;
+}
+
+async function exchangeAuthorizationCode(options: {
+  authBaseUrl: string;
+  code: string;
+  redirectUri: string;
+  verifier: string;
+}) {
+  const { authBaseUrl, code, redirectUri, verifier } = options;
+  return await httpJson<{
+    access_token: string;
+    token_type: string;
+    expires_in: number;
+    id_token?: string;
+    refresh_token?: string;
+  }>(`${authBaseUrl}/v1/token`, {
+    method: 'POST',
+    body: {
+      grant_type: 'authorization_code',
+      code,
+      redirect_uri: redirectUri,
+      code_verifier: verifier,
+    },
+  });
+}
+
+async function persistInstance(options: {
+  instanceName: string;
+  backendBaseUrl: string;
+  clientId: string;
+  clientSecret: string;
+  token: { access_token: string; refresh_token?: string; expires_in: number };
+}) {
+  const { instanceName, backendBaseUrl, clientId, clientSecret, token } =
+    options;
+  const secretStore = await getSecretStore();
+  await withMetadataLock(async () => {
+    const service = `backstage-cli:instance:${instanceName}`;
+    await secretStore.set(service, 'clientSecret', clientSecret);
+    if (token.refresh_token) {
+      await secretStore.set(service, 'refreshToken', token.refresh_token);
+    } else {
+      process.stderr.write(
+        'Warning: No refresh token received. You will need to re-authenticate when the access token expires.\n',
+      );
+    }
+    await upsertInstance({
+      name: instanceName,
+      baseUrl: backendBaseUrl,
+      clientId,
+      accessToken: token.access_token,
+      issuedAt: Date.now(),
+      accessTokenExpiresAt: Date.now() + token.expires_in * 1000,
+    });
+  });
+}
+
+function cryptoRandom(): string {
+  return crypto.randomBytes(16).toString('hex');
+}
+
+// The react-dev-utils/openBrowser breaks the login URL by encoding the URL parameters again
+export function openInBrowser(url: string): void {
+  const platform = process.platform;
+  if (platform === 'darwin') {
+    spawn('open', [url], { detached: true, stdio: 'ignore' }).unref();
+    return;
+  }
+  if (platform === 'win32') {
+    spawn('cmd', ['/c', 'start', '""', url], {
+      detached: true,
+      stdio: 'ignore',
+    }).unref();
+    return;
+  }
+  // linux and others
+  spawn('xdg-open', [url], { detached: true, stdio: 'ignore' }).unref();
```

Comment:
Since these all ultimately call the same function, why not make that more obvious/clear:

```suggestion
  const command = ({ darwin: 'open', win32: 'cmd' }[platform]) ?? 'xdg-open';
  const parameters = platform === 'win32' ? ['/c', 'start', '""', url] : [url];
  
  spawn(command, parameters, { detached: true, stdio: 'ignore' }).unref();
```